## <a href='https://projecteuler.net/problem=44'>44. Pentagon numbers</a>
<a href='https://en.wikipedia.org/wiki/Pentagonal_number'>Pentagonal numbers</a> are generated by the formula, 
$$ P_n= \frac{n(3n−1)}{2} $$ 

The first ten pentagonal numbers are:

$$ 1, 5, 12, 22, 35, 51, 70, 92, 117, 145, ... $$

It can be seen that $P_4 + P_7 = 22 + 70 = 92 = P_8$. However, their difference, $70 − 22 = 48$, is not pentagonal.

Find the pair of pentagonal numbers, 
$$ P_j \text{ & } P_k $$
for which their sum and difference are pentagonal, 

and 
$$ D = |P_k − P_j| $$
is minimised; 

what is the value of D?
___

the maths:  

let:
$$
\begin{aligned}
    P_j &= \frac{3j^2-j}{2} \\
    P_k &= \frac{3k^2-k}{2} \\
    P_j &< P_k
\end{aligned}
$$

since we want their sum $P_m$ and difference $P_n$ are pentagonal:  
- sum
$$
\begin{aligned}
    P_m &= P_j + P_k \\
    \frac{3m^2-m}{2} &= \frac{3j^2-j}{2} + \frac{3k^2-k}{2} \\
    3m^2-m &= 3(j^2 + k^2) - (j+k) 
\end{aligned}
$$
let $M = 3(j^2 + k^2) - (j+k)$, gives an quadratic equation:  
$$
\begin{aligned}
    3m^2 - m - M &= 0 \\
    m &= \frac{1+\sqrt{1+12M}}{6}
\end{aligned}
$$

- difference 
$$
\begin{aligned}
    P_n &= P_k - P_j \\
    \frac{3n^2-n}{2} &= \frac{3k^2-k}{2} - \frac{3j^2-j}{2} \\
    3n^2-n &= 3(k^2 - j^2) - (k-j) 
\end{aligned}
$$
let $N = 3(k^2 - j^2) - (k-j)$, gives an quadratic equation:  
$$
\begin{aligned}
    3n^2 - n - N &= 0 \\
    n &= \frac{1+\sqrt{1+12N}}{6}
\end{aligned}
$$

so for different possible combinations of $j$ and $k$, check if $M$ and $N$ satisfy the quadratic equations and give integer. 

In [1]:
def pentagon(n: int) -> int:
    return n*(3*n-1)//2

def isPentagonal(k: int, generalised: bool=False) -> bool:
    '''
    generalised: check if n is generalised pentagonal number (default False)
    '''
    delta = (1+24*k)**0.5
    if generalised:
        return delta%1 == 0
    else:
        if delta%1 == 0:
            return delta%6 == 5
        else:
            return False

In [2]:
def q44():
    # setup
    k = 1
    found = False
    # since upper bound not known, use while loop
    while not found: 
        # j < k
        for j in range(1, k-1):
            Pk = pentagon(k)
            Pj = pentagon(j)
            Pm = Pk + Pj
            Pn = Pk - Pj
            # check if Pentagonal
            if isPentagonal(Pm) and isPentagonal(Pn):
                found = True
                return print('minimised difference of 2 pentagonal numbers which their sum and difference are pentagonal is %i'%Pn)
        k += 1

In [3]:
%%timeit -n 1 -r 1
q44()

minimised difference of 2 pentagonal numbers which their sum and difference are pentagonal is 5482660
1.75 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


___
further info: 
$$
\begin{aligned}
    P_{1020} &= 1560090 \text{ , } j=1020 \\
    P_{2167} &= 7042750 \text{ , } k=2167 \\
    P_{2395} &= 8602840 \text{ , } m=2395 \\
    P_{1912} &= 5482660 \text{ , } n=1912  
\end{aligned}
$$